sys/crypto: Enable support for AES-192, AES-256

Ollrogge opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This PR adds support for AES-192 and AES-256. The code in /crypto/aes.c did already support AES-192 and AES-256 but the functionality was not made available. This PR adds the additional interfaces to enable AES-192 and AES-256.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
I have added tests for 192 bit and 256 bit keys to the already existing tests for CBC, CTR and ECB mode. The tests have the same origin as the original ones.

I was unable to find official test vectors for 192 bit and 256 bit for CCM and OCB mode. If someone can point me towards test vectors for these modes I can add them as well.

### Issues/PRs references
Fixes #16178, ~~waiting for #16253~~

PeterKietzmann commented:
Fixes #16178

PeterKietzmann commented:
@Ollrogge thanks for your contribution, it is very useful!
> I have added tests for 192 bit and 256 bit keys to the already existing tests for CBC, CTR and ECB mode. The tests have the same origin as the original ones.
- Please point to the origin. Did you lcheck [NIST SP 800-38A ](https://csrc.nist.gov/publications/detail/sp/800-38a/final)?

- Why do you replace `aes_interface` with three interfaces for 128/192/256 bit keys? Their structures only differ in key size, which is configured during `aes_init` anyway, right?
- You have extended `cipher_context_t`, which is fair, to use *aes_* functions directly. However, I'm wondering if this obsoletes [`max_key_size`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/crypto/ciphers.h#L80) in ciphers.h. Do you have opinions on that? I know, the ciphers module is a beast...

PeterKietzmann commented:
BTW: All tests pass on native and nrf51dk and the PR currently adds about 500 Bytes ROM.

Ollrogge commented:
> @Ollrogge thanks for your contribution, it is very useful!
> 
> > I have added tests for 192 bit and 256 bit keys to the already existing tests for CBC, CTR and ECB mode. The tests have the same origin as the original ones.
> 
>     * Very nice! Please point to the origin. Did you check [NIST SP 800-38A ](https://csrc.nist.gov/publications/detail/sp/800-38a/final)?
Yes I did. Sadly it does not contain test vectors for CCM and OCB mode.
> 
>     * Why do you replace `aes_interface` with three interfaces for 128/192/256 bit keys? Their structures only differ in key size, which is configured during `aes_init` anyway, right?
True. By removing [`max_key_size`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/crypto/ciphers.h#L80) they are the same. I did not create a general interface with a name like  `CIPHER_AES` since this would break all code using the `CIPHER_AES_128` interface.

> 
>     * You have extended `cipher_context_t`, which is fair, to use _aes__ functions directly. However, I'm wondering if this obsoletes [`max_key_size`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/crypto/ciphers.h#L80) in ciphers.h. Do you have opinions on that? I know, the ciphers module is a beast...
Yes I think it does. I removed [`max_key_size`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/crypto/ciphers.h#L80). The check in [`ciphers.c`](https://github.com/RIOT-OS/RIOT/blob/e768a85f6225b28ad9347debe36c01bb1f265825/sys/crypto/ciphers.c#L25) is not important I think, since the specific modules check the key size anyways.

PeterKietzmann commented:
> I did not create a general interface with a name like CIPHER_AES since this would break all code using the CIPHER_AES_128 interface

How? Can you explain in a bit more detail? If `aes_interface` was allocated with something like `AES_KEY_SIZE_UNINITIALIZED` and in `aes_init` (which is always called before computing a cipher) you set the keysize for the context (like you did  [here](https://github.com/RIOT-OS/RIOT/pull/16183/commits/0896df7a523d0d71306efa056b148fa73bfda3d3#diff-866e38bdffb92120698ed4452b76c4572f09287b9bcafdbd12a3739bb1ffbb59L821)), what breaks?

PeterKietzmann commented:
> Sadly it does not contain test vectors for CCM and OCB mode.

How about the [Cryptographic Algorithm Validation Program CAVP](https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/cavp-testing-block-cipher-modes#CCM)? 

> The tests have the same origin as the original ones.

What is the origin?

Ollrogge commented:
> 
> How? Can you explain in a bit more detail? If `aes_interface` was allocated with something like `AES_KEY_SIZE_UNINITIALIZED` and in `aes_init` (which is always called before computing a cipher) you set the keysize for the context (like you did [here](https://github.com/RIOT-OS/RIOT/pull/16183/commits/0896df7a523d0d71306efa056b148fa73bfda3d3#diff-866e38bdffb92120698ed4452b76c4572f09287b9bcafdbd12a3739bb1ffbb59L821)), what breaks?

Sorry, I mean the `cipher_id_t` which would break other code in my opinion. Currently it is named [`CIPHER_AES_128`](https://github.com/RIOT-OS/RIOT/blob/e768a85f6225b28ad9347debe36c01bb1f265825/sys/crypto/aes.c#L53) . If I change it to e.g. `CIPHER_AES` code like [this](https://github.com/RIOT-OS/RIOT/blob/e768a85f6225b28ad9347debe36c01bb1f265825/sys/hashes/cmac.c#L52) would break.

I could however remove the interfaces and leave just 1 interface called `aes_interface` and have all ids point to the same interface. I think that is better ?

Ollrogge commented:
> 
> How about the [Cryptographic Algorithm Validation Program CAVP](https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/cavp-testing-block-cipher-modes#CCM)?
> 
Nice ! Thank you, these have cases for 192 and 256 bit keys. I will add these.
> 
> What is the origin?

[this](http://csrc.nist.gov/publications/nistpubs/800-38a/sp800-38a.pdf) is where I took the additional tests for 192 and 256 bit keys for CBC, CTR and ECB mode from.

PeterKietzmann commented:
> I could however remove the interfaces and leave just 1 interface called aes_interface and have all ids point to the same interface. I think that is better ?

Sounds reasonable!

Ollrogge commented:
Hey @PeterKietzmann,

I have added the additional AES-CCM test cases for 192 and 256 bit keys.
Additionally I changed the AES interfaces the way we talked about.

PeterKietzmann commented on sys/crypto/aes.c:
```diff
@@ -45,12 +45,14 @@
  */
 static const cipher_interface_t aes_interface = {
     AES_BLOCK_SIZE,
-    AES_KEY_SIZE,
```
This must be considered as an API change which I would like to avoid here. My [comment](https://github.com/RIOT-OS/RIOT/pull/16183#issuecomment-796800999) was only meant for clarification, not to propose such a change.

PeterKietzmann commented on sys/crypto/aes.c:
```diff
@@ -1037,7 +1038,7 @@ int aes_encrypt(const cipher_context_t *context, const uint8_t *plainBlock,
 
     res = aes_set_encrypt_key((unsigned char *)context->context,
-                              AES_KEY_SIZE * 8, &aeskey);
+                              context->key_size * 8, &aeskey);
```
What if `context->key_size` is not initialized? This happens easily when a user does not utilize the `ciphers` module (see Fortuna RNG). Wouldn't be necessary with the old `cipher_interface_t` struct.

PeterKietzmann commented on sys/crypto/ciphers.c:
```diff
@@ -22,30 +22,24 @@
-
     cipher->interface = cipher_id;
-    return cipher->interface->init(&cipher->context, key, key_size);
+    cipher->context.key_size = key_size;
```
I would rather deal with that in a follow-up PR and focus on `aes_*` here (see comment about API change above)

PeterKietzmann commented on sys/include/crypto/ciphers.h:
```diff
@@ -41,9 +41,7 @@ extern "C" {
-#if defined(MODULE_CRYPTO_3DES)
-    #define CIPHER_MAX_CONTEXT_SIZE 24
-#elif defined(MODULE_CRYPTO_AES)
+#if defined(MODULE_CRYPTO_3DES) || defined(MODULE_CRYPTO_AES)
```
Before, `CIPHER_MAX_CONTEXT_SIZE` was `24` for `MODULE_CRYPTO_3DES`, now is `32`.

-> What was the intention behind this?

PeterKietzmann commented on sys/crypto/aes.c:
```diff
@@ -1305,7 +1306,7 @@ int aes_decrypt(const cipher_context_t *context, const uint8_t *cipherBlock,
 
     res = aes_set_decrypt_key((unsigned char *)context->context,
-                              AES_KEY_SIZE * 8, &aeskey);
+                              context->key_size * 8, &aeskey);
```
Same as above.

PeterKietzmann commented on sys/include/crypto/ciphers.h:
```diff
@@ -65,6 +63,7 @@ extern "C" {
  * @brief   the context for cipher-operations
  */
 typedef struct {
+    uint8_t key_size;                          /**< key size used */
```
See comment above.

PeterKietzmann reviewed (changes requested):
@Ollrogge great improvement on the tests! Still some opens comments, though. Also, there is  a ~500 Byte increase in the `.text` segment that I could not explain at first sight. Can you inspect that? Cosy (see #16083) might be of help.

leandrolanzieri commented:
> @Ollrogge great improvement on the tests! Still some opens comments, though. Also, there is a ~500 Byte increase in the `.text` segment that I could not explain at first sight. Can you inspect that? Cosy (see #16083) might be of help.

I checked the .elf files with [elf_diff](https://github.com/CapeLeidokos/elf_diff). Most of the difference comes from `aes_set_encrypt_key`. The implementation gets bigger. Looking at the code my guess is that, as now it is being called with a non-constant size ([here](https://github.com/RIOT-OS/RIOT/pull/16183/files#diff-866e38bdffb92120698ed4452b76c4572f09287b9bcafdbd12a3739bb1ffbb59R1040-R1041)), many of the conditionals can't be optimized on compile-time.

![elf_diff_report](https://user-images.githubusercontent.com/5381296/111993446-ed32c180-8b16-11eb-9596-71a70af13a24.png)

An option that might work is to conditionally enable support for the different sizes using pesudomodules. After that, one option could be a macro `AES_SIZE(context)`, which would change between constant numbers or `context->key_size` depending on the enable modules. Another one would be to add `IS_USED()` checks on the `if`s withing the function and do the optimization manually.

Ollrogge commented:
Another option @PeterKietzmann and I talked about was to leave the implementation as is and make `AES_KEY_SIZE` configurable via KConfig. This would be sufficient for my use case.

PeterKietzmann commented:
@leandrolanzieri +1

Could you show a short piece of pseudo code to assist implementing it?

leandrolanzieri commented:
I was thinking on defining the pseudomodules on the Makefiles, and then have a macro like:
```C
#if IS_USED(MODULE_AES_128) && !IS_USED(MODULE_AES_192) && !IS_USED(MODULE_AES_256)
#  define KEY_SIZE(ctx) AES_KEY_SIZE_128
#elif !IS_USED(MODULE_AES_128) && IS_USED(MODULE_AES_192) && !IS_USED(MODULE_AES_256)
#  define KEY_SIZE(ctx) AES_KEY_SIZE_192
#elif !IS_USED(MODULE_AES_128) && !IS_USED(MODULE_AES_192) && IS_USED(MODULE_AES_256)
#  define KEY_SIZE(ctx) AES_KEY_SIZE_256
#else
#  define KEY_SIZE(ctx) ctx->keySize
#endif
```

Ollrogge commented:
Hey @leandrolanzieri,
I implemented the handling of the `AES_KEY_SIZE` as you suggested. Using 128 bit keys now only adds 24 Bytes to the .text section.

Additionally I removed all the changes in [ciphers.c](https://github.com/RIOT-OS/RIOT/blob/master/sys/crypto/ciphers.c) as you suggested @PeterKietzmann. 

I hope its okay that I already squashed the commits. My fixups got really messy in the end.

PeterKietzmann commented:
#16253 was merged. Please update this PR as discussed offline.

Ollrogge commented:
I updated the PR based on the changes from [#16253 ](https://github.com/RIOT-OS/RIOT/pull/16253). Using 128 bit keys now adds +8 bytes to the .text section compared to master.

While changing the [CIPHERS_MAX_KEY_SIZE](https://github.com/RIOT-OS/RIOT/blob/c8cb79c3cf3c743fd1ae63c9efd489d4eab13adb/sys/include/crypto/ciphers.h#L34) to be adjusted based on the key sizes used I noticed that it was 20 bytes before. I don't really understand why because 16 bytes should have been sufficient for 128 bit keys ? When using 16 bytes instead of 20 bytes for 128 bit keys the specific tests still pass.

leandrolanzieri commented on sys/include/crypto/aes.h:
```diff
@@ -46,20 +46,35 @@ typedef uint8_t u8;
+/**
+ * @brief AES key size default config
+ */
+#if !IS_USED(MODULE_CRYPTO_AES_128) && !IS_USED(MODULE_CRYPTO_AES_192) && \
+    !IS_USED(MODULE_CRYPTO_AES_256)
+    #define MODULE_CRYPTO_AES_128 1
+#endif
```
The `MODULE_` macros are injected by the build system. It would make sense to raise an `#error` under the condition where no module is being used.

leandrolanzieri commented on sys/include/crypto/ciphers.h:
```diff
@@ -41,10 +53,10 @@ extern "C" {
-#elif defined(MODULE_CRYPTO_AES)
+#if defined(MODULE_CRYPTO_AES)
     #define CIPHER_MAX_CONTEXT_SIZE CIPHERS_MAX_KEY_SIZE
+#elif defined(MODULE_CRYPTO_3DES)
+    #define CIPHER_MAX_CONTEXT_SIZE 24
```
I think 3DES has been removed already, right?

Ollrogge replied:
Oh I didn't notice that. I will remove that define.

It is kinda confusing (atleast to me) that the [documentation](https://doc.riot-os.org/group__sys__crypto.html) still mentions that you can use `crypto_3des` by adding it to `USEMODULE`.

PeterKietzmann replied:
Leftover from #6960

leandrolanzieri commented on makefiles/pseudomodules.inc.mk:
```diff
@@ -298,6 +298,9 @@ PSEUDOMODULES += skald_eddystone
 # define optimized read function of DS18 driver as a pseudo module
 PSEUDOMODULES += ds18_optimized
 
+PSEUDOMODULES += crypto_aes_128
```
Should `crypto_aes_128` be added as a `DEFAULT_MODULE` if `crypto` is used? I think that would match the current expected behaviour.

Ollrogge replied:
Can you give an example of how this would work with `DEFAULT_MODULE` ? I can't find any documentation about it.

Why would this approach be better than my approach of setting `MODULE_CRYPTO_AES_128` if none of the pseudomodules has been specified ?

leandrolanzieri replied:
> Can you give an example of how this would work with DEFAULT_MODULE ? I can't find any documentation about it.

https://github.com/RIOT-OS/RIOT/blob/569e299e63dc7dd132ff528b02b4c1a76d43f563/sys/Makefile.dep#L787-L788
What's added to `DEFAULT_MODULE` can be disabled by `DISABLE_MODULE`.

> Why would this approach be better than my approach of setting `MODULE_CRYPTO_AES_128` if none of the pseudomodules has been specified ?

Because just defining a macro on a specific header file is not the same as activating a module. The build system is in charge of resolving this dependencies, because it can affect other modules, configurations or even bring in other dependencies. The `MODULE_` macros are injected via `riotbuild.h`, so they are available at compilation to all modules, even if they don't include the ciphers header file.

Ollrogge replied:
Okay I see. Thank you !

One more question though: If I add `crypto_aes_128` as `DEFAULT_MODULE` it will always be included. This will remove the optimization we added with `AES_KEY_SIZE(ctx)` if someone wants to use only 192 bit or only 256 bit keys.

leandrolanzieri replied:
As it is added as `DEFAULT_MODULE` it can be disabled via `DISABLE_MODULE`, in case the user only needs AES 256. It's true that the optimization is lost if more than one module is included. But I think it's important to keep the current expected behaviour, meaning AES 128 is available when `crypto` is included.

Maybe some documentation on this would be useful, indicating the available modules and the default configuration.

Ollrogge replied:
Updated the PR. I added documentation about the the behavior at the defines for the `AES_KEY_SIZE_` because this is where I would look for such information.

PeterKietzmann commented:
> I noticed that it was 20 bytes before. I don't really understand why because 16 bytes should have been sufficient

I also stumbled over this and it seems like a leftover or unclean implementation. As indicated by your tests, the four extra bytes have not been used.

leandrolanzieri commented on sys/include/crypto/aes.h:
```diff
@@ -46,20 +46,35 @@ typedef uint8_t u8;
+ * @name AES key sizes
+ *
+ * The default key size is 128 bits. To use a different key size add
+ * USE_MODULE += crypto_aes_192 and/ or USE_MODULE += crypto_aes_256 to
```
```suggestion
 * USEMODULE += crypto_aes_192 and/ or USEMODULE += crypto_aes_256 to
```

leandrolanzieri commented on sys/include/crypto/ciphers.h:
```diff
@@ -56,7 +65,7 @@ extern "C" {
 #define CIPHER_ERR_ENC_FAILED         -5
 #define CIPHER_ERR_DEC_FAILED         -6
-/** Is returned by the cipher_init functions, if the corresponding alogirithm has not been included in the build */
+/** Is returned by the cipher_init functions, if the corresponding algorithm has not been included in the build */
```
Line is longer than 100 chars

Ollrogge commented:
Hey @PeterKietzmann, @leandrolanzieri just pushed my latest fixup.

Is there anything else that needs to changed ?

PeterKietzmann commented on sys/include/crypto/aes.h:
```diff
@@ -46,20 +46,35 @@ typedef uint8_t u8;
+ * an optimized build based on having a static key size.
+ *
+ * @{
+ */
```
- Please move the documentation to the description of the module (top of this file).
- The 128 Bit key size **can** be disabled as an optimization.
- "based on having"? -> based on a static key size. Maybe give this paragraph an other read.

PeterKietzmann commented on sys/crypto/aes.c:
```diff
@@ -49,8 +71,12 @@ static const cipher_interface_t aes_interface = {
     aes_decrypt
 };
+
+/* @deprecated */
```
Please check if this is actually compiled by doxygen. I think it is `/**` for the start. Furthermore, please mention the replacement in the doxygen comment. Have a look at [`CONFIG_ASYMCUTE_BUFSIZE_EXP`](http://doc.riot-os.org/group__net__asymcute__conf.html#ga1947f6d1ba0c46781cb335a6c5bad1d6) as an example.

leandrolanzieri replied:
A removal date is needed, place it after two release cycles, so after 2021.07

Ollrogge replied:
Fixed

PeterKietzmann commented on sys/include/crypto/ciphers.h:
```diff
@@ -23,27 +23,36 @@
- * threedes     needs 24  bytes                           <br>
  * aes          needs CIPHERS_MAX_KEY_SIZE bytes          <br>
  */
-#if defined(MODULE_CRYPTO_3DES)
```
I know it this is picky but can you put the 3DES removal into its own commit?

PeterKietzmann commented on sys/include/crypto/ciphers.h:
```diff
@@ -96,8 +107,10 @@ typedef struct cipher_interface_st {
 
 typedef const cipher_interface_t *cipher_id_t;
 
+/* @deprecated */
```
See comment above. I don't think we need to duplicate the deprecation hint, but you have to check which file is better.  Probably the header file, but `CIPHER_AES` is allocated in aes.c ...

PeterKietzmann commented on tests/sys_crypto/Makefile:
```diff
@@ -2,7 +2,9 @@ include ../Makefile.tests_common
 
 USEMODULE += embunit
 
-USEMODULE += crypto_3des
```
This goes into the  3DES removal commit.

Ollrogge replied:
Done. I hope I separated the commits correctly.

PeterKietzmann commented on tests/sys_crypto/tests-crypto-modes-cbc.c:
```diff
@@ -101,16 +157,88 @@ static void test_decrypt_op(uint8_t *key, uint8_t key_len, uint8_t iv[16],
 {
-    test_encrypt_op(TEST_1_KEY, TEST_1_KEY_LEN, TEST_1_IV, TEST_1_PLAIN,
-                    TEST_1_PLAIN_LEN, TEST_1_CIPHER, TEST_1_CIPHER_LEN);
+    test_encrypt_op_128(TEST_1_KEY, TEST_1_KEY_LEN, TEST_IV, TEST_PLAIN,
```
I find the unspecific naming (eg.,  `TEST_1_KEY_LEN`) confusing, especially since each function call is specific (`test_encrypt_op_128`). Something like `TEST_KEY_LEN_128` would be better. It is probably not the scope of this PR as it would change all test semantics.

PeterKietzmann commented on tests/sys_crypto/tests-crypto-modes-ccm-128.c:
```diff
@@ -0,0 +1,2438 @@
+/*
```
- This file was generated from NIST SP 800-38C with a helper script which is very nice! Would you mind to share the script for documentation? You could paste the code snipped somewhere here in the PR.
- Introducing CCM mode tests should also go into its own commit.

PeterKietzmann reviewed (commented):
- All tests still pass.
- The single-keysize optimization works. A build of a minimal AES-128 encrypt/decrpyt application for the nrf51dk on master:

![master2](https://user-images.githubusercontent.com/7765855/114673931-03433480-9d07-11eb-891c-11a3784e3bd9.png)


on your branch with only few bytes overhead for the additional checks:

![nils2](https://user-images.githubusercontent.com/7765855/114673952-09391580-9d07-11eb-8f61-5b12f7ff591f.png)

Ollrogge commented:
As requested here is the code to parse the testvectors. It is a little specific for the AES_CCM usecase.

```python
import sys
import re
import os

TEST_PREFIX = ""
TEST_GROUP = -1
TEST_NUMBER = -1

def hex_to_arr(inp):
    out = ""
    for i in range(0, len(inp), 2):
        if i % 16 == 0 and i > 0:
            out += "\n"
        out += f"0x{inp[i:i+2]}, "

    return out

def parse_mac_len(f, l):
    global TEST_GROUP
    global TEST_PREFIX

    mac = l.rstrip().split('Tlen = ')[-1][:-1]
    mac = int(mac, 10)

    tmp = f"static const uint8_t TEST_{TEST_PREFIX}_GROUP_{TEST_GROUP}_MAC_LEN = {mac};\n\n"
    f.write(tmp)

def parse_group_header(f, l):
    tmp = l.replace('[', '/* ')
    tmp = tmp.replace(']', ' */')
    f.write(tmp)

    parse_mac_len(f, l)

def parse_key_len(f, n):
    global TEST_PREFIX
    global TEST_NUMBER
    tmp = f"static const size_t TEST_{TEST_PREFIX}_GROUP_{TEST_GROUP}_KEY_LEN = {n};\n\n"
    f.write(tmp)

def parse_nonce_len(b, n):
    global TEST_PREFIX
    global TEST_NUMBER
    tmp = f"static const size_t TEST_{TEST_PREFIX}_{TEST_NUMBER}_NONCE_LEN = {n};\n\n"
    b.append(tmp)

def parse_adata_len(b, n):
    global TEST_PREFIX
    global TEST_NUMBER
    tmp = f"static const size_t TEST_{TEST_PREFIX}_{TEST_NUMBER}_ADATA_LEN = {n};\n\n"
    b.append(tmp)

def parse_ciphertext_len(b, n):
    global TEST_PREFIX
    global TEST_NUMBER
    tmp = f"static const size_t TEST_{TEST_PREFIX}_{TEST_NUMBER}_EXPECTED_LEN = {n};\n\n"
    b.append(tmp)

def parse_input_len(b, n):
    global TEST_PREFIX
    global TEST_NUMBER
    tmp = f"static const size_t TEST_{TEST_PREFIX}_{TEST_NUMBER}_INPUT_LEN = {n};\n\n"
    b.append(tmp)

def parse_key(f, l):
    global TEST_PREFIX
    global TEST_GROUP

    l = l.strip().split('= ')[-1]
    key = f"static const uint8_t TEST_{TEST_PREFIX}_GROUP_{TEST_GROUP}_KEY[] = {{ \n"
    arr =  hex_to_arr(l.rstrip().split('= ')[-1])
    key += arr
    key += "\n};\n"
    f.write(key)

    parse_key_len(f, len(l) // 2)

def parse_test_number(l):
    global TEST_NUMBER
    TEST_NUMBER = int(l.rstrip().split('= ')[-1], 10)

def parse_nonce(b, l):
    global TEST_PREFIX
    global TEST_NUMBER

    l = l.strip().split('= ')[-1]

    nonce = f"static const uint8_t TEST_{TEST_PREFIX}_{TEST_NUMBER}_NONCE[] = {{ \n"
    arr = hex_to_arr(l)
    nonce += arr
    nonce += "\n};\n"
    b.append(nonce)

    if len(l) == 2:
        l = ""

    parse_nonce_len(b, len(l) // 2)

def parse_adata(b, l):
    global TEST_PREFIX
    global TEST_NUMBER

    l = l.strip().split('= ')[-1]

    adata = f"static const uint8_t TEST_{TEST_PREFIX}_{TEST_NUMBER}_ADATA[] = {{ \n"
    arr = hex_to_arr(l)
    adata += arr
    adata += "\n};\n"
    b.append(adata)

    if len(l) == 2:
        l = ""

    parse_adata_len(b, len(l) // 2)

def parse_ciphertext(b, l):
    global TEST_PREFIX
    global TEST_NUMBER

    l = l.strip().split('= ')[-1]

    ciphertext = f"static const uint8_t TEST_{TEST_PREFIX}_{TEST_NUMBER}_EXPECTED[] = {{ \n"
    arr = hex_to_arr(l)
    ciphertext += arr
    ciphertext += "\n};\n"
    b.append(ciphertext)

    if len(l) == 2:
        l = ""

    parse_ciphertext_len(b, len(l) // 2)

def parse_input(b, l):
    global TEST_PREFIX
    global TEST_NUMBER

    l = l.strip().split('= ')[-1]

    inp = f"static const uint8_t TEST_{TEST_PREFIX}_{TEST_NUMBER}_INPUT[] = {{ \n"
    arr = hex_to_arr(l)
    inp += arr
    inp += "\n};\n"
    b.append(inp)

    if len(l) == 2:
        l = ""

    parse_input_len(b, len(l) // 2)

def commit_testcase(f, tc):
    for l in tc:
        f.write(l)

def add_test_calls(f, test_groups, test_nums):
    global TEST_PREFIX
    global TEST_GROUP

    buf = "static void test_crypto_modes_ccm_decrypt2(void)\n"
    buf += "{\n"

    cnt = 0
    for i, group_cnt in enumerate(test_groups):
        for j in range(group_cnt):
            buf += f"do_test_decrypt_op({TEST_PREFIX}, {test_nums[cnt]}, {i});\n"

            cnt += 1

    buf += "}\n"

    buf += "static void test_crypto_modes_ccm_encrypt2(void)\n"
    buf += "{\n"

    cnt = 0
    for i, group_cnt in enumerate(test_groups):
        for j in range(group_cnt):
            buf += f"do_test_encrypt_op({TEST_PREFIX}, {test_nums[cnt]}, {i});\n"

            cnt += 1

    buf += "}\n"

    f.write(buf)

def do_work():
    global TEST_PREFIX
    global TEST_GROUP

    with open(sys.argv[1], "r") as f:
        data = f.readlines()

    TEST_PREFIX = os.path.basename(sys.argv[1]).split('.')[0]

    with open("./test/out.c", "w") as f:
        f.write('''
        #include <limits.h>

        #include <stdlib.h>
        #include <stdio.h>
        #include <string.h>

        #include "embUnit.h"
        #include "crypto/ciphers.h"
        #include "crypto/modes/ccm.h"
        #include "tests-crypto.h"

        static const size_t nonce_and_len_encoding_size = 15;

        ''')

        test_groups = []
        test_nums = []
        test_case = []

        group_cnt = -1

        for l in data:
            if l[0] == '[':
                TEST_GROUP += 1
                parse_group_header(f, l)
                if group_cnt > 0:
                    test_groups.append(group_cnt)

                group_cnt = 0
            elif "Key " in l:
                parse_key(f, l)
            elif "Count " in l:
                test_case = []
                parse_test_number(l)
            elif "Nonce " in l:
                parse_nonce(test_case, l)
            elif "Adata " in l:
                parse_adata(test_case, l)
            elif "CT " in l:
                parse_ciphertext(test_case, l)
            elif "Payload " in l:
                parse_input(test_case, l)
                commit_testcase(f, test_case)

                test_nums.append(TEST_NUMBER)
                group_cnt += 1

        test_groups.append(group_cnt)

        with open("./skeleton.c") as f2:
            f.write(f2.read())

        add_test_calls(f, test_groups, test_nums)

if __name__ == "__main__":
    do_work()
```

PeterKietzmann commented:
ACK. @leandrolanzieri mind do give an ACK too?

leandrolanzieri commented on makefiles/pseudomodules.inc.mk:
```diff
@@ -298,6 +298,9 @@ PSEUDOMODULES += skald_eddystone
 # define optimized read function of DS18 driver as a pseudo module
 PSEUDOMODULES += ds18_optimized
 
+PSEUDOMODULES += crypto_aes_128
```
`crypto_aes` is now `crypto_aes_128`, that should be changed in the `Makefile.dep`s.

leandrolanzieri replied:
Also https://github.com/RIOT-OS/RIOT/blob/master/sys/crypto/Kconfig needs to be updated. Some tests are failing because of this. I think something like the following should work:
```difff
diff --git a/sys/crypto/Kconfig b/sys/crypto/Kconfig
index fa070b10c1..ce99ad759a 100644
--- a/sys/crypto/Kconfig
+++ b/sys/crypto/Kconfig
@@ -19,18 +19,24 @@ choice
         The common Crypto block ciphers API has multiple implementations. Choose
         one of the following.
 
-config MODULE_CRYPTO_AES
+config CRYPTO_AES
     bool "AES"
     select MODULE_CRYPTO
 
-config MODULE_CRYPTO_3DES
-    bool "3DES (deprecated)"
-    select MODULE_CRYPTO
-
 endchoice
 
 menu "Crypto AES options"
-depends on MODULE_CRYPTO_AES
+depends on CRYPTO_AES
+
+config MODULE_CRYPTO_AES_128
+    bool "AES-128"
+    default y
+
+config MODULE_CRYPTO_AES_192
+    bool "AES-192"
+
+config MODULE_CRYPTO_AES_256
+    bool "AES-256"
 
 config MODULE_CRYPTO_AES_PRECALCULATED
     bool "Pre-calculate T tables"
```

Ollrogge replied:
Added the configuration. How do I actually test this ? The configs depend on TEST_KCONFIG. Setting this variable to 1 in my Makefile results in errors.

leandrolanzieri replied:
Something like the following should show you which modules are selected when using Kconfig
`env BOARD=samr21-xpro TEST_KCONFIG=1 make info-modules`.

leandrolanzieri replied:
It seems that there are still some occurrences of `crypto_aes` in some Makefile.dep files

Ollrogge replied:
Missed some. Should all be gone now. I also removed the `crypto_%`  pseudomodules now.

leandrolanzieri replied:
Thanks

leandrolanzieri commented:
@Ollrogge thanks for updating the makefiles! It looks like just a couple of boards are overflowing after splitting the crypto tests. For those you can simply add them to the `BOARD_INSUFFICIENT_MEMORY` list, like so: https://github.com/RIOT-OS/RIOT/blob/master/tests/pkg_libcoap/Makefile.ci Also, you will need to rebase as there is a conflict with `sys/Makefile.dep`.

I think the test application names are not really descriptive, could you try to use a more precise name for them? Also, it would be nice to add some `README.md` in the apps to describe shortly what is tested on both applications, and perhaps a reference to where the test vectors come from.

Ollrogge commented:
> @Ollrogge thanks for updating the makefiles! It looks like just a couple of boards are overflowing after splitting the crypto tests. For those you can simply add them to the `BOARD_INSUFFICIENT_MEMORY` list, like so: https://github.com/RIOT-OS/RIOT/blob/master/tests/pkg_libcoap/Makefile.ci Also, you will need to rebase as there is a conflict with `sys/Makefile.dep`.
> 
> I think the test application names are not really descriptive, could you try to use a more precise name for them? Also, it would be nice to add some `README.md` in the apps to describe shortly what is tested on both applications, and perhaps a reference to where the test vectors come from.

Rebased my changes onto upstream/master and added the overflowing boards to `BOARD_INSUFFICIENT_MEMORY`. I also added the README's to the test applications. I renamed `sys_crypto` test application to `sys_crypto_aes_ccm` because this test application is only testing `AES_CCM`. I don't really know how to rename the `sys_crypto` test application because it literally tests all parts of the crypto module and therefore the general name seems appropriate to me.

Since my latest fixup commit the Murdock test is failing completely. I don't really understand why ? Can you tell me what I did wrong ?

leandrolanzieri commented on sys/Makefile.dep:
```diff
@@ -105,9 +113,13 @@ ifneq (,$(filter fuzzing,$(USEMODULE)))
-# include GNRC dependencies
-ifneq (,$(filter gnrc% %gnrc,$(USEMODULE)))
-  include $(RIOTBASE)/sys/net/gnrc/Makefile.dep
+ifneq (,$(filter gnrc_lorawan,$(USEMODULE)))
+  USEMODULE += xtimer
+  USEMODULE += random
+  USEMODULE += hashes
+  USEMODULE += crypto_aes_128
+  USEMODULE += netdev_layer
+  USEMODULE += gnrc_nettype_lorawan
```
I think this is why the tests are failing now, all the dependency resolution of GNRC was removed, perhaps introduced while rebasing.

leandrolanzieri replied:
Due to #16268 being merged. All GNRC-specify dependencies have been moved to their own file.

Ollrogge replied:
Thanks ! Fixed.

leandrolanzieri commented on sys/crypto/aes.c:
```diff
@@ -49,7 +71,9 @@ static const cipher_interface_t aes_interface = {
 };
+
 const cipher_id_t CIPHER_AES_128 = &aes_interface;
+const cipher_id_t CIPHER_AES = &aes_interface;
```
Doing a `git grep CIPHER_AES_128`  shows that there are still many uses of the deprecated `cipher_id`, including the crypto documentation. Those should be updated to use `CIPHER_AES` instead.

Ollrogge replied:
Fixed !

leandrolanzieri reviewed (commented):
I added one more comment regarding the dependency resolution. Otherwise it looks good. Feel free to squash after applying the fix @Ollrogge

Ollrogge commented on sys/Makefile.dep:
```diff
@@ -42,25 +42,37 @@ ifneq (,$(filter i2c_scan,$(USEMODULE)))
 
-ifneq (,$(filter crypto_aes_%,$(USEMODULE)))
-  USEMODULE += crypto_aes
+ifneq (,$(filter crypto_aes_unroll,$(USEMODULE)))
+  USEMODULE += crypto_aes_128
+endif
+
+ifneq (,$(filter crypto_aes_precalculated,$(USEMODULE)))
+  USEMODULE += crypto_aes_128
```
These lines actually don't matter at all I think since `crypto_aes_128` is a default module if `crypto` is used and crypto is included based on `crypto_%`.

PeterKietzmann commented:
My ACK holds, but I haven't followed the latest changes that deal with the size of the tests. @leandrolanzieri wanna merge?

leandrolanzieri commented:
I'm running tests of the affected modules, just to be on the safe side.

fjmolinas commented:
> I'm running tests of the affected modules, just to be on the safe side.

Tested OpenWSN, nodes join fine (this is the part where crypto is used):

```
18:25:00 SUCCESS 5de6 [IEEE802154E] synchronized at slotOffset 0
18:25:00 SUCCESS 67eb [IEEE802154E] synchronized at slotOffset 0
18:25:07 INFO 5de6 [C6T] sending CJOIN request
18:25:08 VERBOSE New node: ae8d:fee1:60c7:cb8a. Creating new OSCORE context in /builds/tmp/RIOT/bin/oscore_context_ae8dfee160c7cb8a.json.
18:25:08 VERBOSE received JRC join request
18:25:08 VERBOSE 5de6 [PACKETFUNCTIONS] copy packet content to big packet (pkt len 1300 > max len 125)
18:25:08 VERBOSE 5de6 [PACKETFUNCTIONS] copy packet content to small packet (pkt len 33 < max len 125)
18:25:08 SUCCESS 5de6 [C6T] node joined
18:25:09 INFO 67eb [C6T] sending CJOIN request
18:25:10 VERBOSE New node: ae8d:fee1:e02c:e22d. Creating new OSCORE context in /builds/tmp/RIOT/bin/oscore_context_ae8dfee1e02ce22d.json.
18:25:10 VERBOSE received JRC join request
18:25:10 VERBOSE 67eb [PACKETFUNCTIONS] copy packet content to big packet (pkt len 1300 > max len 125)
18:25:10 VERBOSE 67eb [PACKETFUNCTIONS] copy packet content to small packet (pkt len 33 < max len 125)
18:25:10 SUCCESS 67eb [C6T] node joined
18:25:22 INFO 67eb [SIXTOP] sending a 6top request
18:25:22 INFO 5de6 [SIXTOP] sending a 6top request
18:25:23 SUCCESS 67eb [SIXTOP] sixtop return code RC_SUCCESS at sixtop state WAIT_ADDRESPONSE
18:25:24 SUCCESS 5de6 [SIXTOP] sixtop return code RC_SUCCESS at sixtop state WAIT_ADDRESPONSE
18:25:26 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - parents:
           bbbb:0:0:0:ae8d:fee1:208c:870d
        - children:
18:25:35 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - parents:
           bbbb:0:0:0:ae8d:fee1:208c:870d
        - children:
18:25:46 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - parents:
           bbbb:0:0:0:ae8d:fee1:208c:870d
        - children:

```

leandrolanzieri commented:
I have tested the following, which work as expected:
- `semtech-loramac` (using test app)
- `gnrc_lorawan` (using example)
- `ieee802154_security` (by adding it to `gnrc_networking`, as suggested #15150)

leandrolanzieri reviewed (approved):
Changes look good, the feature works as expected while keeping the default behaviour and code-size because of the optimizations in place. Tests have been greatly improved.

ACK! Thanks @Ollrogge for the contribution!

leandrolanzieri commented:
And GO!

benpicco commented on sys/include/crypto/aes.h:
```diff
@@ -46,20 +54,26 @@ typedef uint8_t u8;
-};
-
-typedef struct aes_key_st AES_KEY;
+} AES_KEY;
```
Why not?
```suggestion
} aes_key_t;
```
